src/PostgREST/Auth/JwtCache.hs

```diff
@@ -1,99 +1,112 @@
-
--- | Used to retrieve and insert JWT to JWT Cache
-lookupJwtCache :: JwtCacheState -> ByteString -> Int -> IO (Either Error AuthResult) -> UTCTime -> IO (Either Error AuthResult)
-lookupJwtCache JwtCacheState{jwtCache, purgeCache} token maxLifetime parseJwt utc = do
-  checkCache <- C.lookup jwtCache token
-  authResult <- maybe parseJwt (pure . Right) checkCache
-
-  case (authResult,checkCache) of
-    -- From comment:
-    -- https://github.com/PostgREST/postgrest/pull/3801#discussion_r1857987914
-    --
-    -- We purge expired cache entries on a cache miss
-    -- The reasoning is that:
-    --
-    -- 1. We expect it to be rare (otherwise there is no point of the cache)
-    -- 2. It makes sure the cache is not growing (as inserting new entries
-    --    does garbage collection)
-    -- 3. Since this is time expiration based cache there is no real risk of
-    --    starvation - sooner or later we are going to have a cache miss.
-
-    (Right res, Nothing) -> do -- cache miss
-
-      let timeSpec = getTimeSpec res maxLifetime utc
-
-      -- insert new cache entry
-      C.insert' jwtCache (Just timeSpec) token res
-
-      -- Execute IO action to purge the cache
-      -- It is assumed this action returns immidiately
-      -- so that request processing is not blocked.
-      purgeCache
-
-    _                    -> pure ()
-
-  return authResult
-
--- Used to extract JWT exp claim and add to JWT Cache
-getTimeSpec :: AuthResult -> Int -> UTCTime -> TimeSpec
-getTimeSpec res maxLifetime utc = do
-  let expireJSON = KM.lookup "exp" (authClaims res)
-      utcToSecs = floor . nominalDiffTimeToSeconds . utcTimeToPOSIXSeconds
-      sciToInt = fromMaybe 0 . Sci.toBoundedInteger
-  case expireJSON of
-    Just (JSON.Number seconds) -> TimeSpec (sciToInt seconds - utcToSecs utc) 0
-    _                          -> TimeSpec (fromIntegral maxLifetime :: Int64) 0
-
--- | Empty the cache (done when the config is reloaded)
-emptyCache :: JwtCacheState -> IO ()
-emptyCache JwtCacheState{jwtCache} = C.purge jwtCache
+newJwtCache :: AppConfig -> ObservationHandler -> IO JwtCache
+newJwtCache AppConfig{configJWKS, configJwtCacheMaxEntries} observationHandler = do
+  maybe (pure JwtNoJwks) initCache configJWKS
+  where
+    initCache key = createCache key configJwtCacheMaxEntries
+
+    createCache key maxSize = do
+          maxSizeTVar <- newTVarIO maxSize
+          JwtCache key maxSizeTVar <$>
+            -- select cachingErrors or notCachingErrors
+            notCachingErrors (readTVar maxSizeTVar) key
+
+    cachingErrors :: STM Int -> JwkSet -> IO (SC.Cache IO ByteString (Either Error JSON.Object))
+    cachingErrors maxSize key = SC.cacheIO (SC.CacheConfig maxSize
+            (runExceptT . parseAndDecodeClaims key)
+            (observationHandler . JwtCacheLookup) -- lookup metrics
+            (const . const $ observationHandler JwtCacheEviction) -- evictions metrics
+            alwaysValid) -- no invalidation for now
```

Comment:
We should now delete this and remove the top `{-# OPTIONS_GHC -Wno-unused-local-binds #-}` right?